### PR TITLE
ROX-26381: Column mgmt workload CVE overview

### PR DESCRIPTION
**Pull request by dvail**
### Description

Implements column management for the Workload CVE overview page using PatternFly's `ColumnManagementModal` component and the `useManagedColumns()` hook in #12904 .

**Note:** This PR is dependent on #12904 and is awaiting team discussion on the naming conventions for `tableId` strings, but structurally is ready for review.

### Follow ups

1. Implementation in the Workload CVE detail pages
2. Integration tests
3. If time, beautification of the toolbar layout on smaller screen resolutions

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

Visit Workload CVEs and see the presence of the "Manage columns" button in the toolbar, and a count of the number of columns that are enabled.
image

Clicking on the button opens a modal listing the columns that are available for configuration:
image

Uncheck a column and save changes:
image
image

Test the remaining columns and verify that each column is hidden individually.

Do a hard refresh of the page and see that column state has persisted:
image

Test that the "Select all" and "Restore to defaults" options in the modal reset the state:
image

Test that applying the state causes the columns to re-appear:
image

Verify that "Images" and "Deployment" tabs have their own independent set of columns:
image
image

Verify that each column works correctly individually for the Images and Deployments tabs.

### CODE REVIEW

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx [hunk 1]
```diff
@@ -74,28 +105,46 @@ function DeploymentsTable({
     filteredSeverities,
     showCveDetailFields,
     onClearFilters,
+    tableConfig,
 }: DeploymentsTableProps) {
     const vulnerabilityState = useVulnerabilityState();
-    const colSpan = showCveDetailFields ? 6 : 5;
+    const getVisibilityClass = generateVisibilityFor(tableConfig);
```
Comment (pedrottimark) [inline ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:111]: Maybe an impression from scrolling back and forth for review more than following the pattern:

Compare `tableConfig` here to `columnState` argument in function definition. How about?
* `generateVisibilityForColumns` although dangling prepositions do not bother me in general
* `columnVisibilityState` in call and definition

Or feel free to suggest another alternative that aligns the above with the following related names:
* `getVisibilityClass`
* `getHiddenColumnCount`

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx [hunk 2]
```diff
@@ -138,7 +187,7 @@ function DeploymentsTable({
                                         </Link>
                                     </Td>
                                     {showCveDetailFields && (
-                                        <Td>
+                                        <Td className={getVisibilityClass('cvesBySeverity')}>
```
Comment (pedrottimark) [inline ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:190]: Thinking about naming compared to other names related to table cells caused me to notice absence of `dataLabel` prop, except for **Deployment** column (if you expand unchanged code above).
deployments_dataLabel

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx [hunk 3]
```diff
@@ -11,11 +11,41 @@ import TooltipTh from 'Components/TooltipTh';
...
+    generateVisibilityFor,
+    getHiddenColumnCount,
+    ManagedColumns,
+} from 'hooks/useManagedColumns';
 import { getWorkloadEntityPagePath } from '../../utils/searchUtils';
 import SeverityCountLabels from '../../components/SeverityCountLabels';
 import { VulnerabilitySeverityLabel } from '../../types';
 import useVulnerabilityState from '../hooks/useVulnerabilityState';
 
+export const tableId = 'workloadCvesDeploymentOverviewTable';
```
Comment (pedrottimark) [inline ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:24]: Understanding that this is placeholder pending brainstorming by the team:
1. `WorkloadCves` prefix corresponding to folder (as discussed in chat)
2. What are pro and con of  `DeploymentsTable` file name versus`DeploymentOverviewTable` as suffix?
    * One side of coin: reuse previous decisions?
    * Other side of coin: does more specific name suggest what file name would have been in hindsight?
        If yes, to limit changed files, I do not consider file renaming as necessary to preseve pattern.

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx (hunk 2 above)
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:190]: If you decide to add `dataLabel` prop while you are editing `Td` elements, also ImagesTable.tsx file.

Comment (dvail) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:190]: Good call, will do.

(Good candidate for a lint rule?)

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx (hunk 3 above)
Comment (dvail) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:24]: Hmm... the file names of these tables have been on the back burner for a while. If we go with the folder prefix + file name combination, I'd be in favor of just doing the rename now so long as it doesn't result in a large enough change that git detects it as a new file. If so, I could do the rename in a follow up.

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx (hunk 2 above)
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx:190]: Indeed, wrote first draft of `Td-dataLabel` which finds 51 errors in 19 files.

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx (hunk 3 above)
Comment (dvail) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentsTable.tsx]: I went ahead with this change and the file renaming. Although we will still have this on the agenda for the team to collab on, this is an improvement over what was previously in the PR.

File: ui/apps/platform/src/Components/ColumnManagementButton.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                        newColumns.map(({ key, isShown }) => [key, isShown ?? false])
+                    );
+                    setColumns(nextState);
+                }}
+                isOpen={isOpen}
+                onClose={() => setOpen(false)}
+            />
+            <Button
+                onClick={() => setOpen(true)}
+                variant={ButtonVariant.secondary}
```
Comment (pedrottimark) [inline ui/apps/platform/src/Components/ColumnManagementButton.tsx:36]: Nuance of PatternFly for awareness. I trust your judgement which alternative.
* `ButtonVariant` and its family look like support for application specific wrapper classes:
    https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/Button/Button.tsx#L8-L18
* `variant` property has inline TypeScript string enumeration as source of truth:
    https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/Button/Button.tsx#L79

File: ui/apps/platform/src/Components/ColumnManagementButton.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        <>
+            <ColumnManagementModal
+                appliedColumns={Object.values(columns)}
+                applyColumns={(newColumns) => {
+                    const nextState = Object.fromEntries(
+                        newColumns.map(({ key, isShown }) => [key, isShown ?? false])
+                    );
+                    setColumns(nextState);
+                }}
+                isOpen={isOpen}
```
Comment (pedrottimark) [inline ui/apps/platform/src/Components/ColumnManagementButton.tsx:31]: What are pro and con:
* Unconditional rendering of instance with `isOpen` prop.
* Conditional rendering like `{isOpen && …}` with `isOpen={true}` prop

File: ui/apps/platform/src/Components/TooltipTh.tsx [hunk 6]
```diff
@@ -3,16 +3,17 @@ import { Tooltip } from '@patternfly/react-core';
...
     tooltip: string;
 };
 
 // this is to ensure that the tooltip always shows up on hover since
 // the tooltip prop on Th only shows when the header is truncated
-function TooltipTh({ children, tooltip, sort }: TooltipThProps) {
+function TooltipTh({ className, children, tooltip, sort }: TooltipThProps) {
     return (
-        <Th sort={sort || undefined}>
+        <Th className={className} sort={sort || undefined}>
```
[the remark this answers is not in the document]
Comment (dvail) [reply ui/apps/platform/src/Components/TooltipTh.tsx:16]: Personally I wouldn't mind a lint or TS rule that bans `||` except when used with a preceding boolean value, but I'm not sure how feasible that is. There isn't any difference in this particular case but it is easy to get comfortable and use it with a string or numeric type and introduce edge case bugs.

File: ui/apps/platform/src/Components/ColumnManagementButton.tsx (hunk 5 above)
Comment (dvail) [reply ui/apps/platform/src/Components/ColumnManagementButton.tsx:31]: In this case, I don't think there is any pro to conditional rendering. An `isOpen` prop of `false` will cause the PatternFly component to be rendered as `null`, and the conditional rendering on our end would only increase the JSX indent and add a couple of lines of code.

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/CVEsTable.tsx [hunk 7]
```diff
@@ -139,15 +165,19 @@ function CVEsTable({
...
+    const getVisibilityClass = generateVisibilityFor(tableConfig);
+    const hiddenColumnCount = getHiddenColumnCount(tableConfig);
 
     const colSpan =
         6 +
         (canSelectRows ? 1 : 0) +
         (createTableActions ? 1 : 0) +
-        (showExceptionDetailsLink ? 1 : 0);
+        (showExceptionDetailsLink ? 1 : 0) +
+        -hiddenColumnCount;
```
[the remark this answers is not in the document]
Comment (sachaudh) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/CVEsTable.tsx:180]: Using variables may also help a bit here:

```
const baseColSpan = 6;
const selectRowColSpan = canSelectRows ? 1 : 0;
const tableActionsColSpan = createTableActions ? 1 : 0;
const exceptionDetailsColSpan = showExceptionDetailsLink ? 1 : 0;

const colSpan = sum([
    baseColSpan,
    selectRowColSpan,
    tableActionsColSpan,
    exceptionDetailsColSpan,
    -hiddenColumnCount
]);
```

**Review by pedrottimark — APPROVED** [review]

**Review by sachaudh — APPROVED** [review]
Overall looks good to me. Thanks for driving this feature 👍🏼

File: ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentOverviewTable.tsx [hunk 8]
```diff
@@ -58,44 +88,63 @@ export type Deployment = {
...
     showCveDetailFields,
     onClearFilters,
-}: DeploymentsTableProps) {
+    columnVisibilityState,
+}: DeploymentOverviewTableProps) {
     const vulnerabilityState = useVulnerabilityState();
-    const colSpan = showCveDetailFields ? 6 : 5;
+    const getVisibilityClass = generateVisibilityForColumns(columnVisibilityState);
+    const hiddenColumnCount = getHiddenColumnCount(columnVisibilityState);
+    const colSpan = 5 + (showCveDetailFields ? 1 : 0) + -hiddenColumnCount;
```
[the remark this answers is not in the document]
Comment (dvail) [reply ui/apps/platform/src/Containers/Vulnerabilities/WorkloadCves/Tables/DeploymentOverviewTable.tsx:113]: Yes, in the case of feature flags we would have to move this into the body of the component and calculate dynamically. We'd lose a bit of type-safety with the helper functions but it isn't a huge deal and should be easily caught by testing.

For not allowing columns to be hidden, we would just omit that column from the `defaultColumns` object. There is also the notion of `isUntoggleAble` that is built into the PatternFly component, but I don't see a use case for that at the moment where showing an untoggleable column in the modal would be preferable.

**Pull request merged**